File: preprocessing/partitioning/Gambit2Metis/gambit2metis5.f90
!>
!! @file
!! This file is part of SeisSol.
!!
!! @section LICENSE
!! Copyright (c) SeisSol Group
!! All rights reserved.
!! 
!! Redistribution and use in source and binary forms, with or without
!! modification, are permitted provided that the following conditions are met:
!! 
!! 1. Redistributions of source code must retain the above copyright notice,
!!    this list of conditions and the following disclaimer.
!! 
!! 2. Redistributions in binary form must reproduce the above copyright notice,
!!    this list of conditions and the following disclaimer in the documentation
!!    and/or other materials provided with the distribution.
!! 
!! 3. Neither the name of the copyright holder nor the names of its
!!    contributors may be used to endorse or promote products derived from this
!!    software without specific prior written permission.
!! 
!! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
!! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
!! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
!! ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
!! LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
!! CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
!! SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
!! INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
!! CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
!! ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
!! POSSIBILITY OF SUCH DAMAGE.

program gambit2metis5
  
  use mesh_mod
  implicit none

  character(:),allocatable :: infile,metisfile,filename,file
  character(len=512) :: arg
  type(meshtype)     :: mesh
  integer            :: n,np

  call get_command_argument(1,arg)
  infile=trim(adjustl(arg));n=len(infile)
  call get_command_argument(2,arg)
  read(arg,*)np
  call mesh%read_gambit(infile)
  filename=get_filename(infile);filename=filename(1:len(filename)-4)
  call do_partitioning_external(mesh,np,filename)

contains

  subroutine do_partitioning_external(mesh,np,filename)
    use ifport, only:system
    implicit none
    type(meshtype) :: mesh
    character(:),allocatable :: metisfile
    character(len=*) :: filename
    integer :: np
    integer :: i,iun
    metisfile=filename//'.met'  
    open(newunit=iun,file=metisfile,action='write')
    write(iun,*)mesh%nelem,1
    do i = 1,mesh%nelem
       write(iun,*)mesh%cells(i)%nodes
    end do
    close(iun)
    i=system('m2gmetis '//filename//'.met '//filename//'.met.dgraph ')
    i=system('gpmetis '//filename//'.met.dgraph '//int2str(np))
    !clean up of files
    i=system('rm '//filename//'.met')
    i=system('rm '//filename//'.met.dgraph')
    i=system('mv '//filename//'.met.dgraph.part.'//int2str(np)//' '//filename//'.met.epart.'//int2str(np))
  end subroutine do_partitioning_external

  function get_filename(filestring,mext)
    implicit none
    character(len=*) :: filestring
    character(:), allocatable :: get_filename
    integer, optional :: mext
    integer :: i
    do i=len(filestring),1,-1
       if(filestring(i:i)=='/')exit
    end do
    get_filename=trim(adjustl(filestring(i+1:len(filestring))))
  end function get_filename

  function int2str(i)
    integer :: i
    character(:), allocatable :: int2str
    character(len=20):: dum
    write(dum,*)i
    int2str=trim(adjustl(dum))
  end function int2str

!!$  subroutine do_partitioning_api
!!$
!!$    use metis_mod
!!$    USE, INTRINSIC :: ISO_C_BINDING
!!$
!!$    integer :: objval,icount,numflag,ncommon
!!$    integer, allocatable, dimension(:) :: cellnodes,cellindices,e_mpiid,n_mpiid,eptr,eind
!!$    integer, allocatable, dimension(:),target ::   xadj,adjncy
!!$    !integer,pointer     ::   pxadj(:)=> NULL(),padjncy(:)=> NULL()
!!$    integer, pointer :: vwgt(:)  => NULL()
!!$    integer, pointer :: vsize(:) => NULL()
!!$    integer, pointer :: mopts(:) => NULL()
!!$    real(kind(1.d0)), pointer :: tpwgts(:) => NULL()
!!$    integer::METIS_PartMeshDual
!!$  
!!$    allocate(cellnodes(1:sum(mesh%cells%nvert)))
!!$    allocate(cellindices(1:mesh%nelem+1))
!!$    icount=1
!!$    cellindices(1)=1
!!$    do j = 1,mesh%nelem
!!$       cellindices(j+1)=cellindices(j)+mesh%cells(j)%nvert
!!$       do i = 1,mesh%cells(j)%nvert
!!$          cellnodes(icount)=mesh%cells(j)%nodes(i)
!!$          icount =icount+1
!!$       enddo
!!$       if(j<10)then
!!$          write(*,*)mesh%cells(j)%nodes,cellindices(j),cellindices(j+1)-1
!!$          write(*,*)cellnodes(cellindices(j):cellindices(j+1)-1)
!!$       endif
!!$    enddo
!!$    
!!$    allocate(eind(0:sum(mesh%cells%nvert)-1))
!!$    allocate(eptr(0:mesh%nelem))
!!$    
!!$    eind=cellnodes
!!$    eptr=cellindices
!!$    
!!$    allocate(e_mpiid(mesh%nelem),n_mpiid(mesh%numnp))
!!$    write(*,*)'   partitioning mesh into ',np,' parts'
!!$    write(*,*)'   ...  calling metis_partmeshdual ...'
!!$    call METIS_SetDefaultOptions(metis_options);
!!$    call mymetis_default_options();mopts=>null()!metis_options
!!$    ncommon=1
!!$    numflag=1
!!$    i=METIS_PartMeshDual(mesh%nelem,mesh%numnp,eptr,eind, & 
!!$         & vwgt,vsize,ncommon,np,tpwgts,metis_options,objval,e_mpiid,n_mpiid)
!!$    write(*,*)'   ...          metis_partmeshdual done!'
!!$    write(*,*)'objval',objval,i
!!$    
!!$    file=filename//'.met.epart2.'//int2str(np)
!!$    open(newunit=iun,file=file)
!!$    do i = 1,size(e_mpiid,1)
!!$       write(iun,'(I1)')e_mpiid(i)
!!$    enddo;close(iun);
!!$
!!$  file=filename//'.met.epart3.'//int2str(np)
!!$  open(newunit=iun,file=file)  
!!$  do i = 1,size(n_mpiid,1)
!!$     write(iun,'(I1)')n_mpiid(i)
!!$  enddo;close(iun);
!!$
!!$  end subroutine do_partitioning_api

end program gambit2metis5

File: preprocessing/partitioning/gambit2metis5.m
%%
% @file
% This file is part of SeisSol.
%
% @author Martin Kaeser (martin.kaeser AT geophysik.uni-muenchen.de, http://www.geophysik.uni-muenchen.de/Members/kaeser)
% @author Sebastian Rettenberger (rettenbs AT in.tum.de, http://www5.in.tum.de/wiki/index.php/Sebastian_Rettenberger,_M.Sc.)
%
% @section LICENSE
% Copyright (c) 2008-2012, SeisSol Group
% All rights reserved.
% 
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are met:
% 
% 1. Redistributions of source code must retain the above copyright notice,
%    this list of conditions and the following disclaimer.
% 
% 2. Redistributions in binary form must reproduce the above copyright notice,
%    this list of conditions and the following disclaimer in the documentation
%    and/or other materials provided with the distribution.
% 
% 3. Neither the name of the copyright holder nor the names of its
%    contributors may be used to endorse or promote products derived from this
%    software without specific prior written permission.
% 
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
% AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
% IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
% ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
% LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
% CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
% SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
% INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
% CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
% ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.

home;
disp(' ')
disp('     %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%')
disp('     %%                                                             %%')
disp('     %% Gambit2Metis converts a GAMBIT-meshfile of                  %%')
disp('     %% triangles, quadrilaterals, tetrahedrons or hexahedrons      %%')
disp('     %% stored as "filename.neu"                                    %%')
disp('     %% into a METIS-file and calls METIS for mesh partitioning.    %%')
disp('     %% METIS mesh partitioner:                                     %%')
disp('     %% http://www-users.cs.umn.edu/~karypis/metis/metis/index.html %%')
disp('     %%                                                             %%')
disp('     %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%')
disp(' '),disp(' ')
clear, close all;
filename = input('     Specify GAMBIT-filename:   ','s');

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%        Read Gambit Data 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
t = cputime;

fid   = fopen([filename,'.neu']);
    disp(' ')
    disp('-----------------------------------------------------------------------------------')
    disp(sprintf('\t Reading data from: \t\t%s',[filename,'.neu']));
    for i= 1:6
        junk  = fgetl(fid);
    end
    tmp    = fscanf(fid,'%i',[5,1]); 
    NUMNP  = tmp(1);                     %Number of Vertices
    NELEM  = tmp(2);                     %Number of Elements 
    %NGRPS  = tmp(3);                    %Number of Element Groups
    %NBSETS = tmp(4);                    %Number of Boundary Condition Sets
    NDFCD  = tmp(5)                     %Number of Dimensions (2 or 3)

    switch NDFCD                         %Choose between 2-D or 3-D mesh input
        case 2     %2D-meshes
            disp(sprintf('\t Reading \t\t\t\t\t%i vertices',NUMNP));
            disp(sprintf('\t Reading \t\t\t\t\t%i elements',NELEM));
            junk  = fgetl(fid); junk  = fgetl(fid); junk  = fgetl(fid);
            X     = fscanf(fid,'%g',[3,NUMNP]);                            %Read Vertices
            junk  = fgetl(fid); junk  = fgetl(fid); junk  = fgetl(fid);
            %
            first        = fscanf(fid,'%i',[3,1])
            element_type = first(2)
            clear X; 
            switch element_type                                            %Read Elements
                case 3  %triangular mesh
                    elem1 = fscanf(fid,'%i',[3,1]);
                    tri   = fscanf(fid,'%i',[6,NELEM-1]);
                    tri   = [elem1, tri(4:6,:)];
                    disp(sprintf('\t Read all triangles successfully!'));
                    ncommon = '2';

                case 2  %quadrilateral mesh
                    elem1 = fscanf(fid,'%i',[4,1]);
                    quad  = fscanf(fid,'%i',[7,NELEM-1]);
                    quad  = [elem1, quad(4:7,:)];
                    disp(sprintf('\t Read all quadrilaterals successfully!'));
                    ncommon = '2';
            end
                        
        case 3     %3D-meshes
            disp(sprintf('\t Reading \t\t\t\t\t%i vertices',NUMNP));
            disp(sprintf('\t Reading \t\t\t\t\t%i tetrahedrons',NELEM));
            junk  = fgetl(fid); junk  = fgetl(fid); junk  = fgetl(fid);
            X     = fscanf(fid,'%g',[4,NUMNP]);                            %Read Vertices
            junk  = fgetl(fid); junk  = fgetl(fid); junk  = fgetl(fid);
            %
            first        = fscanf(fid,'%i',[3,1]);
            element_type = first(2);
            clear X; 
            switch element_type                                            %Read Elements
                case 6  %tetrahedral mesh
                    elem1 = fscanf(fid,'%i',[4,1]);
                    tetra = fscanf(fid,'%i',[7,NELEM-1]);
                    tetra = [elem1, tetra(4:7,:)];
                    disp(sprintf('\t Read all tetrahedrons successfully!'));
                    ncommon = '3';

                case 4  %hexahedral mesh
                    elem1 = fscanf(fid,'%i',[8,1]);
                    hexa  = fscanf(fid,'%i',[11,NELEM-1]);
                    hexa  = [elem1, hexa(4:11,:)];
                    disp(sprintf('\t Read all hexahedrons successfully!'));
                    ncommon = '4';
            end
    
        otherwise
            disp(sprintf('\n\t ##ERROR: The GAMBIT input file shows a problem\n\n ##'))
            return;
            
    end
fclose(fid);
disp('-----------------------------------------------------------------------------------')

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%        Writing Metis Data 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
disp(sprintf('\t Write METIS-file: \t\t\t%s',[filename,'.met']));

fid = fopen([filename,'.met'],'w');
    fprintf(fid,'%12i \n',NELEM);
    switch NDFCD
        case 2
           switch element_type
               case 3
                  fprintf(fid,'%12i %12i %12i \n',tri);
               case 2
                  fprintf(fid,'%12i %12i %12i %12i\n',quad);
           end
            
        case 3
            switch element_type
               case 6
                  fprintf(fid,'%12i %12i %12i %12i \n',tetra);
               case 4
                  fprintf(fid,'%12i %12i %12i %12i %12i %12i %12i %12i\n',hexa);
            end
    end
fclose(fid);
disp('-----------------------------------------------------------------------------------')
disp(sprintf('\n\t Conversion finished successfully!  (%g CPU sec)\n',cputime-t));
disp('-----------------------------------------------------------------------------------')
disp(' ')
pause(0.2);
proc = input('     Specify number of processor of the partition:   ','s');
disp('-----------------------------------------------------------------------------------')
disp(sprintf('\n\t METIS partition starts!\n'));
disp('-----------------------------------------------------------------------------------')
disp(' ')
%eval(['!partnmesh ',filename,num2str(i),'.met ',proc])
system(['m2gmetis -ncommon=',ncommon,' ',filename,'.met ',filename,'.met.dgraph'])
system(['gpmetis -ptype=rb ',filename,'.met.dgraph ',proc])
pause(0.5);
%clean up of files
delete([filename,'.met'])
delete([filename,'.met.dgraph'])    
movefile([filename,'.met.dgraph.part.',proc],[filename,'.met.epart.',proc])

disp('-----------------------------------------------------------------------------------')
disp(sprintf('\t METIS partition done!'));
disp('-----------------------------------------------------------------------------------')
disp(' ')
disp('-----------------------------------------------------------------------------------')
disp(sprintf('\t Wrote final METIS-file: \t\t%s',[filename,'.met.epart.',num2str(proc)]));
disp('-----------------------------------------------------------------------------------')
disp(' ')

